### PR TITLE
feat: trun workflows

### LINKED ISSUE #328: ensure examples run examples in Deno+puppeteer in CI
We need a minimal way to test that simple examples run in the browser. This should take the form of a check that appropriately block merging upon fails.

Related

- #437

### PR DESCRIPTION
**Pull request by ryanleecode**
Example testing workflows with https://github.com/paritytech/trun. Resolves #328

### CODE REVIEW

File: .github/actions/trun/action.yml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    description: run files in a browser
+  importMap:
+    description: import map
+  concurrency:
+    description: concurrency
+  ignore:
+    description: ignore file
+runs:
+  using: "docker"
+  image: "Dockerfile"
```
Comment (harrysolovay) [inline .github/actions/trun/action.yml:17]: This setup is only necessary for the browser example workflow, correct? If so, should we forgo the action definition altogether and inline these instructions in the browser example workflow?

Comment (ryanleecode) [reply .github/actions/trun/action.yml:17]: should be possible

Comment (ryanleecode) [reply .github/actions/trun/action.yml:17]: Nah i dont think it works. dockerfiles only work in actions not workflows

Comment (tjjfvi) [reply .github/actions/trun/action.yml:17]: You can specify `container:` on a job – https://docs.github.com/en/actions/using-jobs/running-jobs-in-a-container#example-running-a-job-within-a-container

**Comment by ryanleecode** [comment]
```
pathInfo {
  vCapi: undefined,
  generatorId: "frame",
  providerId: "dev",
  target: "polkadot",
  vRuntime: "v0.9.39-1",
  filePath: "types/sp_runtime/multiaddress.js"
}
Downloading polkadot/v0.9.39-1/x86_64-unknown-linux-gnu
Download https://deno.land/std@0.181.0/bytes/mod.ts
Download https://deno.land/std@0.181.0/bytes/ends_with.ts
Download https://deno.land/std@0.181.0/bytes/equals.ts
Download https://deno.land/std@0.181.0/bytes/includes_needle.ts
Download https://deno.land/std@0.181.0/bytes/index_of_needle.ts
Download https://deno.land/std@0.181.0/bytes/last_index_of_needle.ts
Download https://deno.land/std@0.181.0/bytes/repeat.ts
Download https://deno.land/std@0.181.0/bytes/starts_with.ts
port 43221
READY error ConnectionRefused: Connection refused (os error 111)
    at async Object.connect (ext:deno_net/01_net.js:466:55)
    at async ready (file:///__w/capi/capi/util/port.ts:11:26)
    at async file:///__w/capi/capi/providers/frame/FrameBinProvider.ts:25:11
    at async PolkadotDevProvider.connect (file:///__w/capi/capi/providers/frame/FrameProxyProvider.ts:60:33)
    at async file:///__w/capi/capi/providers/frame/FrameProvider.ts:96:26
    at async withSignal (file:///__w/capi/capi/util/withSignal.ts:4:12)
    at async file:///__w/capi/capi/providers/frame/FrameProvider.ts:83:29
    at async FsCache._getRaw (file:///__w/capi/capi/util/cache/fs.ts:16:23)
    at async file:///__w/capi/capi/providers/frame/FrameProvider.ts:80:19
    at async Promise.all (index 0)
```

I suspect the binary is just crashing.

**Comment by ryanleecode** [comment]
polkadot binary failing with 

```
Error: 
   0: Failed to get runtime version: cannot deserialize module: UnknownOpcode(192)

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
2023-03-29 00:15:28 Cannot create a runtime error=Other("cannot deserialize module: UnknownOpcode(192)")
```

**Comment by ryanleecode** [comment]
@tjjfvi what rust version are we using to compile the binaries?

somebody ran into a similar issue

https://substrate.stackexchange.com/questions/7714/cannot-run-substrate-on-a-fresh-macbook-m2

**Comment by ryanleecode** [comment]
marking this ready for review as it _should_ just work out of the box as soon as the capi binaries are fixed

**Review by tjjfvi — CHANGES_REQUESTED** [review]
- Can these be incorporated into the new `checks.yml` workflow, rather than being separate?
- Can you add some tasks to `deno.jsonc` for trun, and use them in these workflows?
- Does deno need the container?

**Comment by ryanleecode** [comment]
Blocked by https://github.com/paritytech/trun/issues/5

**Comment by ryanleecode** [comment]
@tjjfvi @harrysolovay regular deno is working but browser is still having issues. Want to just merge the deno part for now?

File: .trunignore [hunk 2]
```diff
@@ -0,0 +1,3 @@
+**/xcm/asset_teleportation.eg.ts
+**/smoldot.eg.ts
+**/ink/*.eg.ts
```
Comment (tjjfvi) [inline .trunignore:3]:
```suggestion
examples/xcm/asset_teleportation.eg.ts
examples/smoldot.eg.ts
examples/ink/*.eg.ts
```

File: deno.jsonc [hunk 3]
```diff
@@ -33,7 +33,7 @@
     "udd": "deno run -A _tasks/udd.ts",
     "dnt": "deno task run _tasks/dnt.ts",
     "test": "deno task capi -- deno test -A -L=info --ignore=target --parallel -r=http://localhost:4646/",
-    "test:examples": "deno task test _tasks/test_examples.ts",
+    "test:examples:deno": "deno task capi -- deno run -A -r=http://localhost:4646/ https://raw.githubusercontent.com/paritytech/trun/v0.1.0-beta.1/main.ts --include \"*/**/*.eg.ts\" --import-map import_map.json --concurrency 1 --reload http://localhost:4646/",
```
Comment (tjjfvi) [inline deno.jsonc:36]:
```suggestion
    "test:examples:deno": "deno task capi -- deno run -A -r=http://localhost:4646/ https://raw.githubusercontent.com/paritytech/trun/v0.1.0-beta.1/main.ts --include \"**/*.eg.ts\" --import-map import_map.json --concurrency 1 --reload http://localhost:4646/",
```

File: .github/workflows/checks.yml [hunk 4]
```diff
@@ -84,3 +84,27 @@ jobs:
             target/capi/**/_chainName
           key: ${{ runner.os }}-capi-metadata-${{ hashFiles('import_map.json') }}
       - run: deno task test
+  test-examples-deno:
+    name: Test Examples (Deno)
```
Comment (tjjfvi) [inline .github/workflows/checks.yml:88]:
```suggestion
  examples-deno:
    name: Examples (Deno)
```

**Review by harrysolovay — DISMISSED** [review]
This marks a meaningful improvement to our testing. Great work on this one @ryanleecode.

**Review by tjjfvi — APPROVED** [review]

**Comment by ryanleecode** [comment]
3 mins 28 seconds LFG

**Review by harrysolovay — APPROVED** [review]

**Review by tjjfvi — DISMISSED** [review]

**Review by tjjfvi — APPROVED** [review]

**Pull request merged**